reframe/frontend/executors/__init__.py

```diff
@@ -22,14 +23,33 @@ class TestCase:
     def __init__(self, check, partition, environ):
         self.__check_orig = check
         self.__check = copy.deepcopy(check)
-        self.__environ = copy.deepcopy(environ)
         self.__partition = copy.deepcopy(partition)
+        self.__environ = copy.deepcopy(environ)
+        self.__check._case = weakref.ref(self)
+        self.__deps = []
 
     def __iter__(self):
         # Allow unpacking a test case with a single liner:
         #       c, p, e = case
         return iter([self.__check, self.__partition, self.__environ])
 
+    def __hash__(self):
+        return (hash(self.check.name) ^
```

Comment:
Maybe just return the hash of a tuple consisting of the corresponding strings.

Reply:
I think this one is more efficient.

Reply:
I won't change this; I don't see the advantage of what you propose.